lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,748 @@
+		}
+	}
+
+	fn receive_tx_add_output(mut self, serial_id: u64, output: TxOut) -> InteractiveTxStateMachineResult<Negotiating> {
+		// The receiving node:
+		//  - MUST fail the negotiation if:
+		//     - the serial_id has the wrong parity
+		if !self.is_valid_counterparty_serial_id(serial_id) {
+			return self.abort_negotiation(AbortReason::IncorrectSerialIdParity);
+		}
+
+		self.context.received_tx_add_output_count += 1;
+		if self.context.received_tx_add_output_count > MAX_RECEIVED_TX_ADD_OUTPUT_COUNT {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//     - if has received 4096 `tx_add_output` messages during this negotiation
+			return self.abort_negotiation(AbortReason::ReceivedTooManyTxAddOutputs);
+		}
+
+		if output.value < output.script_pubkey.dust_value().to_sat() {
+			// The receiving node:
+			// - MUST fail the negotiation if:
+			//		- the sats amount is less than the dust_limit
+			return self.abort_negotiation(AbortReason::ExceededDustLimit);
+		}
+		if output.value > TOTAL_BITCOIN_SUPPLY_SATOSHIS {
+			// The receiving node:
+			// - MUST fail the negotiation if:
+			//		- the sats amount is greater than 2,100,000,000,000,000 (TOTAL_BITCOIN_SUPPLY_SATOSHIS)
+			return self.abort_negotiation(AbortReason::ExceededMaximumSatsAllowed);
+		}
+
+		// The receiving node:
+		//   - MUST accept P2WSH, P2WPKH, P2TR scripts
+		//   - MAY fail the negotiation if script is non-standard
+		if !output.script_pubkey.is_v0_p2wpkh() && !output.script_pubkey.is_v0_p2wsh() &&
+			!output.script_pubkey.is_v1_p2tr()
+		{
+			return self.abort_negotiation(AbortReason::InvalidOutputScript);
+		}
+
+		if let None = self.context.outputs.insert(serial_id, output) {
+			Ok(InteractiveTxStateMachine { context: self.context, state: Negotiating {} })
+		} else {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the `serial_id` is already included in the transaction
+			self.abort_negotiation(AbortReason::DuplicateSerialId)
+		}
+	}
+
+	fn receive_tx_remove_output(mut self, serial_id: SerialId) -> InteractiveTxStateMachineResult<Negotiating> {
+		if !self.is_valid_counterparty_serial_id(serial_id) {
+			return self.abort_negotiation(AbortReason::IncorrectSerialIdParity);
+		}
+
+		if let Some(output) = self.context.outputs.remove(&serial_id) {
+			Ok(InteractiveTxStateMachine { context: self.context, state: Negotiating {} })
+		} else {
+			self.abort_negotiation(AbortReason::SerialIdUnknown)
+		}
+	}
+
+	fn send_tx_add_input(mut self, serial_id: u64, input: TxIn, prevout: TxOut) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.inputs.insert(
+			serial_id,
+			TxInputWithPrevOutput {
+				input: input,
+				prev_output: prevout
+			}
+		);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_add_output(mut self, serial_id: SerialId, output: TxOut) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.insert(serial_id, output);
```

Comment:
I expect checking of the current state; is that check present somewhere else?

Reply:
This PR is currently very stale. We've been working off a different branch that has these check. I'll push them up to this branch later.